s3fs/core.py

```diff
@@ -461,12 +461,25 @@ def put_tags(self, path, tags):
         path: str
             Existing key to attach tags to
         tags: dict str, str
-            Tags to apply. Will over-write any existing tags.
+            Tags to apply.
+        mode:
+            One of 'o' or 'm'
+            'o': Will over-write any existing tags.
+            'm': Will merge in new tags with existing tags.  Incurs two remote calls.
         """
         bucket, key = split_path(path)
-        tag = {'TagSet': [
-                   {'Key': k, 'Value': v} for k, v in tags.items()]
-               }
+
+        if mode == 'm':
+            existing_tags = self.get_tags(path=path)
+            merged_tags = existing_tags.copy()
+            merged_tags.update(tags)
+            new_tags = [{'Key': k, 'Value': v} for k, v in merged_tags.items()]
+        elif mode == 'o':
+            new_tags = [{'Key': k, 'Value': v} for k, v in tags.items()]
+        else:
+            raise NotImplementedError("Mode must be {'o', 'm'}, not %s" % mode)
```

Comment:
Probably a ValueError - there are no other values of mode that we intend to implement.